File: 2015/09/09.py

from itertools import permutations
from string import ascii_uppercase

with open('./input.txt', 'r') as file:
    problem_input = [line.split() for line in file.read().splitlines()]

name_map, distances = {}, {}
indexer = (x for x in ascii_uppercase)

for line in problem_input:
    if line[0] not in name_map.keys():
        name_map[line[0]] = next(indexer)
    if line[2] not in name_map.keys():
        name_map[line[2]] = next(indexer)

for line in problem_input:
    distances[name_map[line[0]] + name_map[line[2]]] = int(line[-1])

paths = {''.join(p) : 0 for p in set(permutations(set(''.join(distances.keys()))))}

for key in paths.keys():
    for k in (key[i:i+2] for i in range(0, len(key)-1, 1)):
        try:
            paths[key] += distances[k]
        except KeyError:
            paths[key] += distances[k[::-1]]

print(sorted(list(paths.values()))[0])
print(sorted(list(paths.values()))[-1])

# 176.7 ms ±  46.4 ms (1000 runs, 100 warmup) - Python 3.10.5 (conda-forge)

File: 2015/09/09.cpp
#include <iostream>
#include <fstream>
#include <cctype>
#include <array>
#include <vector>
#include <map>
#include <functional>
#include <algorithm>


char indexer (void);
unsigned int factorial(unsigned int);
std::vector<std::string> permute (std::string);


int main (void) {

    std::string data;
    std::ifstream infile;

    std::map<std::string, char> name_map;
    std::string to_permute;
    infile.open("./input.txt");
    while (infile.peek() != EOF) {
        infile >> data;
        if (data == "to" || data.length() == 1 || std::isdigit(data[0]))
            continue;
        if (name_map.find(data) == name_map.end()) {
            char idx {indexer()};
            to_permute += idx;
            name_map[data] = idx;
        }
    }
    infile.close();

    std::string key;
    std::map<std::string, unsigned> distances;
    infile.open("./input.txt");
    while (infile.peek() != EOF) {
        infile >> data;
        if (data == "to" || data[0] == '=')
            continue;
        if (!std::isdigit(data[0])) {
            key += name_map[data];
        }
        if (std::isdigit(data[0]) && key.length() == 2) {
            distances.insert({ key, std::stoi(data) });
            key = "";
        }
    }
    infile.close();

    auto permutations {permute(to_permute)};
    std::map<std::string, int> paths {};
    for (auto &el : permutations) {
        paths.insert({ el, 0 });
    }

    for (auto &item : paths) {
        for (int i {}; i <= item.first.length(); i++) {
            std::string k {item.first.substr(i, 2)};
            if (k.length() == 2) {
                if (distances.find(k) != distances.end()) {
                    paths[item.first] += distances[k];
                } else {
                    std::reverse(k.begin(), k.end());
                    paths[item.first] += distances[k];
                }
            }
        }
    }

    int min, max;
    min = max = paths[to_permute];
    for (auto &item : paths) {
        if (item.second < min) {
            min = item.second;
        }
        if (item.second > max) {
            max = item.second;
        }
    }
    std::cout << min << '\n' << max << std::endl;

}


char indexer (void) {
    static int ch {'A'-1};
    if (ch < 'Z') ch++;
    else ch = 'A';
    return static_cast<char>(ch);
}

unsigned factorial(unsigned num) {
    return (num == 1 || num == 0) ? 1: num * factorial(num - 1);
}

std::vector<std::string> permute (std::string str) {
    std::vector<std::string> permutations {};
    unsigned len {factorial(str.length())};
    permutations.reserve(len);
    do {
        permutations.push_back(str);
    } while (std::next_permutation(str.begin(), str.end()));
    return permutations;
}

// 45.2 ms ± 13.4 ms (10k runs, 1k warmup) - G++ 12.1.1: g++ -O3 -std=C++20